File: euxis-runtime/src/runtime/validator.py
# SPDX-License-Identifier: AGPL-3.0-or-later

from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path
from typing import Iterable


class RuntimeValidationError(ValueError):
    """Raised when runtime assets fail structural validation."""


@dataclass(frozen=True)
class RuntimeValidator:
    """Validate runtime package data/config artifacts."""

    runtime_root: Path

    def ensure_required_files(self) -> None:
        required = [
            self.runtime_root / "README.md",
            self.runtime_root / "config" / "etx-settings.json",
            self.runtime_root / "data" / "perf" / "metrics.jsonl",
            self.runtime_root / "data" / "lifecycle" / "transitions.jsonl",
            self.runtime_root / "data" / "manifests" / "performance-optimization.json",
        ]
        missing = [str(p) for p in required if not p.exists()]
        if missing:
            raise RuntimeValidationError(f"Missing runtime assets: {', '.join(missing)}")

    def _read_jsonl(self, path: Path) -> list[dict]:
        records: list[dict] = []
        for lineno, raw in enumerate(path.read_text(encoding="utf-8").splitlines(), start=1):
            if not raw.strip():
                continue
            try:
                obj = json.loads(raw)
            except json.JSONDecodeError as exc:
                raise RuntimeValidationError(f"{path}:{lineno}: invalid jsonl record") from exc
            if not isinstance(obj, dict):
                raise RuntimeValidationError(f"{path}:{lineno}: record is not an object")
            records.append(obj)
        if not records:
            raise RuntimeValidationError(f"{path}: no records")
        return records

    def _assert_keys(self, records: Iterable[dict], required: set[str], context: str) -> None:
        for idx, record in enumerate(records):
            missing = required.difference(record.keys())
            if missing:
                ordered = ", ".join(sorted(missing))
                raise RuntimeValidationError(f"{context}[{idx}] missing keys: {ordered}")

    def validate_perf_metrics(self) -> None:
        path = self.runtime_root / "data" / "perf" / "metrics.jsonl"
        records = self._read_jsonl(path)
        self._assert_keys(records, {"ts", "op", "agent", "ms"}, str(path))
        for idx, record in enumerate(records):
            ms = record.get("ms")
            if not isinstance(ms, int | float) or ms < 0:
                raise RuntimeValidationError(f"{path}[{idx}] invalid ms value")

    def validate_lifecycle_transitions(self) -> None:
        path = self.runtime_root / "data" / "lifecycle" / "transitions.jsonl"
        records = self._read_jsonl(path)
        self._assert_keys(records, {"ts", "agent", "state", "session"}, str(path))

    def validate_lifecycle_state_files(self) -> None:
        lifecycle_dir = self.runtime_root / "data" / "lifecycle"
        state_files = sorted(lifecycle_dir.glob("*.state"))
        if not state_files:
            raise RuntimeValidationError(f"{lifecycle_dir}: no .state files found")
        for path in state_files:
            raw = path.read_text(encoding="utf-8").strip()
            if not raw:
                raise RuntimeValidationError(f"{path}: empty state file")

    def validate_manifest(self) -> None:
        path = self.runtime_root / "data" / "manifests" / "performance-optimization.json"
        try:
            data = json.loads(path.read_text(encoding="utf-8"))
        except json.JSONDecodeError as exc:
            raise RuntimeValidationError(f"{path}: invalid json") from exc
        if not isinstance(data, dict):
            raise RuntimeValidationError(f"{path}: manifest must be an object")

    def validate(self) -> None:
        self.ensure_required_files()
        self.validate_perf_metrics()
        self.validate_lifecycle_transitions()
        self.validate_lifecycle_state_files()
        self.validate_manifest()


def validate_runtime_layout(runtime_root: str | Path) -> None:
    RuntimeValidator(Path(runtime_root)).validate()

File: pkg/runtime/src/validator.cpp
#include "euxis/runtime/validator.hpp"

#include <algorithm>
#include <fstream>
#include <sstream>

namespace euxis::runtime {

RuntimeValidator::RuntimeValidator(std::filesystem::path runtime_root)
    : root_(std::move(runtime_root)) {}

void RuntimeValidator::validate() const {
    ensure_required_files();
    validate_perf_metrics();
    validate_lifecycle_transitions();
    validate_lifecycle_state_files();
    validate_manifest();
}

void RuntimeValidator::ensure_required_files() const {
    std::vector<std::filesystem::path> required{
        root_ / "README.md",
        root_ / "config" / "etx-settings.json",
        root_ / "data" / "perf" / "metrics.jsonl",
        root_ / "data" / "lifecycle" / "transitions.jsonl",
        root_ / "data" / "manifests" / "performance-optimization.json",
    };

    std::vector<std::string> missing;
    for (const auto& p : required) {
        if (!std::filesystem::exists(p)) {
            missing.push_back(p.string());
        }
    }
    if (!missing.empty()) {
        std::string msg = "Missing runtime assets: ";
        for (size_t i = 0; i < missing.size(); ++i) {
            if (i > 0) msg += ", ";
            msg += missing[i];
        }
        throw RuntimeValidationError(msg);
    }
}

auto RuntimeValidator::read_jsonl(const std::filesystem::path& path) const
    -> std::vector<nlohmann::json> {
    std::ifstream file(path);
    if (!file) throw RuntimeValidationError("Cannot open: " + path.string());

    std::vector<nlohmann::json> records;
    std::string line;
    int lineno = 0;
    while (std::getline(file, line)) {
        ++lineno;
        // Skip blank lines
        if (line.find_first_not_of(" \t\r\n") == std::string::npos) continue;
        try {
            auto obj = nlohmann::json::parse(line);
            if (!obj.is_object()) {
                throw RuntimeValidationError(
                    path.string() + ":" + std::to_string(lineno) +
                    ": record is not an object");
            }
            records.push_back(std::move(obj));
        } catch (const nlohmann::json::parse_error&) {
            throw RuntimeValidationError(
                path.string() + ":" + std::to_string(lineno) +
                ": invalid jsonl record");
        }
    }
    if (records.empty()) {
        throw RuntimeValidationError(path.string() + ": no records");
    }
    return records;
}

void RuntimeValidator::assert_keys(const std::vector<nlohmann::json>& records,
                                   const std::vector<std::string>& required,
                                   const std::string& context) const {
    for (size_t idx = 0; idx < records.size(); ++idx) {
        std::vector<std::string> missing;
        for (const auto& key : required) {
            if (!records[idx].contains(key)) {
                missing.push_back(key);
            }
        }
        if (!missing.empty()) {
            std::ranges::sort(missing);
            std::string keys_str;
            for (size_t i = 0; i < missing.size(); ++i) {
                if (i > 0) keys_str += ", ";
                keys_str += missing[i];
            }
            throw RuntimeValidationError(
                context + "[" + std::to_string(idx) + "] missing keys: " + keys_str);
        }
    }
}

void RuntimeValidator::validate_perf_metrics() const {
    auto path = root_ / "data" / "perf" / "metrics.jsonl";
    auto records = read_jsonl(path);
    assert_keys(records, {"ts", "op", "agent", "ms"}, path.string());

    for (size_t idx = 0; idx < records.size(); ++idx) {
        const auto& ms = records[idx]["ms"];
        if (!ms.is_number() || ms.get<double>() < 0) {
            throw RuntimeValidationError(
                path.string() + "[" + std::to_string(idx) + "] invalid ms value");
        }
    }
}

void RuntimeValidator::validate_lifecycle_transitions() const {
    auto path = root_ / "data" / "lifecycle" / "transitions.jsonl";
    auto records = read_jsonl(path);
    assert_keys(records, {"ts", "agent", "state", "session"}, path.string());
}

void RuntimeValidator::validate_lifecycle_state_files() const {
    auto lifecycle_dir = root_ / "data" / "lifecycle";
    std::vector<std::filesystem::path> state_files;
    for (const auto& entry : std::filesystem::directory_iterator(lifecycle_dir)) {
        if (entry.path().extension() == ".state") {
            state_files.push_back(entry.path());
        }
    }
    if (state_files.empty()) {
        throw RuntimeValidationError(
            lifecycle_dir.string() + ": no .state files found");
    }
    std::ranges::sort(state_files);
    for (const auto& path : state_files) {
        std::ifstream file(path);
        std::string content((std::istreambuf_iterator<char>(file)),
                            std::istreambuf_iterator<char>());
        // Trim
        auto start = content.find_first_not_of(" \t\r\n");
        if (start == std::string::npos) {
            throw RuntimeValidationError(path.string() + ": empty state file");
        }
    }
}

void RuntimeValidator::validate_manifest() const {
    auto path = root_ / "data" / "manifests" / "performance-optimization.json";
    std::ifstream file(path);
    if (!file) throw RuntimeValidationError("Cannot open: " + path.string());

    std::string content((std::istreambuf_iterator<char>(file)),
                        std::istreambuf_iterator<char>());
    try {
        auto data = nlohmann::json::parse(content);
        if (!data.is_object()) {
            throw RuntimeValidationError(
                path.string() + ": manifest must be an object");
        }
    } catch (const nlohmann::json::parse_error&) {
        throw RuntimeValidationError(path.string() + ": invalid json");
    }
}

void validate_runtime_layout(const std::filesystem::path& runtime_root) {
    RuntimeValidator(runtime_root).validate();
}

} // namespace euxis::runtime
